src/gfn/samplers.py

```diff
@@ -264,3 +263,324 @@ def sample_trajectories(
+            non_exit_valid_actions = valid_actions[~valid_actions.is_exit]
+
+            # Using all non-initial states, calculate the backward policy, and the logprobs
+            # of those actions.
+            if new_trajectories.conditioning is not None:
+                # We need to index the conditioning vector to broadcast over the states.
+                cond_dim = (-1,) * len(new_trajectories.conditioning.shape)
+                traj_len = new_trajectories.states.tensor.shape[0]
+                masked_cond = new_trajectories.conditioning.unsqueeze(0).expand(
+                    (traj_len,) + cond_dim
+                )[~new_trajectories.states.is_sink_state][
+                    ~valid_states.is_initial_state
+                ]
+
+                # Pass all valid states, i.e., non-sink states, except the initial state.
+                with has_conditioning_exception_handler(
+                    "pb", self.backward_sampler.estimator
+                ):
+                    estimator_outputs = self.backward_sampler.estimator(
+                        non_initial_valid_states, masked_cond
+                    )
+            else:
+                # Pass all valid states, i.e., non-sink states, except the initial state.
+                with no_conditioning_exception_handler(
+                    "pb", self.backward_sampler.estimator
+                ):
+                    estimator_outputs = self.backward_sampler.estimator(
+                        non_initial_valid_states
+                    )
+
+            valid_log_pb_actions = (
+                self.backward_sampler.estimator.to_probability_distribution(
+                    non_initial_valid_states, estimator_outputs
+                ).log_prob(non_exit_valid_actions.tensor)
+            )
+
+            log_pb_new_trajectories = torch.full_like(
+                new_trajectories.actions.tensor[..., 0],
+                fill_value=0.0,
+                dtype=torch.float,
+            )
+            log_pb_new_trajectories_slice = torch.full_like(
+                valid_actions.tensor[..., 0], fill_value=0.0, dtype=torch.float
+            )
+            log_pb_new_trajectories_slice[~valid_actions.is_exit] = valid_log_pb_actions
+            log_pb_new_trajectories[
+                ~new_trajectories.actions.is_dummy
+            ] = log_pb_new_trajectories_slice
+
+            # TODO: Implement Metropolis-Hastings acceptance criterion.
+            # p(x->s'->x') = p_B(x->s')p_F(s'->x')
+            # p(x'->s'->x) = p_B(x'->s')p_F(s'->x)
+            # The acceptance ratio is
+            # min(1, R(x')p(x->s'->x') / R(x)p(x'->s'-> x))
+            # Note that
+            # p(x->s'->x') / p(x'->s'-> x))
+            # = p_B(x->s')p_F(s'->x') / p_B(x'->s')p_F(s'->x)
+            # = p_B(x->s'->s0)p_F(s0->s'->x') / p_B(x'->s'->s0)p_F(s0->s'->x)
+            # = p_B(tau|x)p_F(tau') / p_B(tau'|x')p_F(tau)
```

Comment:
This is how Metropolis-Hastings works.
In the last line, we need both log_pfs and log_pbs of each trajectory.